### ISSUE TITLE
New API Proposal

### ISSUE THREAD

**Issue by Pyrolistical** [hyperapp/router#4]
Love Hyperapp, but I found the API for this router to be a tad odd. I feel its violating the principle of least surprise.

I don't like how it overload the `view` property as people are used it it being a function, and now its not. Also why are the routes an array when it could be simply an object?

# New API Proposal
```
app({
  routes: {
    "/": (state, actions) =>
      <div>
        <h1>Home</h1>
        <a href="#about">About</a>
      </div>
    ,
    "/:route": (state, actions) =>
      <div>
        <h1>About</h1>
        <a href="#home">Home</a>
      </div>
  },
  mixins: [Router]
})
```

1. Leave `view` alone and even throw an error if `view` is defined when Router mixin is used. Tell the user `view` is expected to be empty and configure `routes`
2. `routes` object instead of array. There will always be more than one route, so the property needed to be plural

**Comment by jbucaran** [comment]
> @Pyrolistical I don't like how it overload the view property as people are used it it being a function, and now its not. 

Yes, I couldn't agree more and that's about to change very soon in when I fix the router for hyperapp@0.10.0.

> Also why are the routes an array when it could be simply an object?

There is a reason for this one. It used to be an object, just like the example you have given. But that makes the router indeterministic due to how objects behave in JavaScript.

If the view is treated as a key/route (function) object, the study the following example:

```
localhost:8080/documents
```
```js
{
  "/documents": 0,
  "/:foo": 1
}
```
In the above example, the matched view may be 0 or 1 depending on which route key is checked first.

## See
- https://github.com/hyperapp/hyperapp/pull/214

**Comment by Pyrolistical** [comment]
That's cool. I minor suggestions to make the API UX slightly easier to use:
```
app({
  routes: [
    "/", state => <h1>Hi.</h1>,
    "*", state => <h1>404</h1>,
  ],
  mixins: [Router]
})
```

I would still add a routes property and not expose the implementation detail of setting view in order to use the Router. I would also make routes take a long list of alternating strings and functions to reduce the syntax. You can always error if the list length is odd or if it doesn't exactly alternate between string and function.
